R/expect-shape.R

```diff
@@ -0,0 +1,57 @@
+#' Does code return an object with the specified shape?
+#'
+#' This is a generalization of [expect_length()] to test the "shape" of
+#'   more general objects like data.frames, matrices, and arrays.
+#'
+#' @seealso [expect_length()] to specifically make assertions about the
+#'   [length()] of a vector.
+#' @inheritParams expect_that
+#' @param ... Ignored.
+#' @param length Expected [length()] of `object`.
+#' @param nrow Expected [nrow()] of `object`.
+#' @param ncol Expected [ncol()] of `object`.
+#' @param dim Expected [dim()] of `object`.
+#' @family expectations
+#' @export
+#' @examples
+expect_shape = function(object, ..., length, nrow, ncol, dim) {
+  # absent, not present, due to '!' operator precedence
+  n_absent <- missing(length) + missing(nrow) + missing(ncol) + missing(dim)
+  if (n_absent != 3L) {
+    cli::cli_abort(
+      "Exactly one of {.arg length}, {.arg nrow}, {.arg ncol}, or {.arg dim} must be provided."
+    )
+  }
+
+  # Re-use expect_length() to ensure they stay in sync.
+  if (!missing(length)) {
+    return(expect_length(object, length))
+  }
+
+  act <- quasi_label(enquo(object), arg = "object")
+  # need base:: qualification or we might trigger an error for missing(dim)
+  dim_object <- base::dim(object)
+
+  if (!missing(nrow)) {
```

Comment:
We might want to first check `length(dim_object)` since I think `expect_shape(1, nrow = 1)` will return a confusing error. Or `expect_shape(array(1), ncol = 1)`?

Reply:
Great call, added some examples & fixed up the code a fair amount